ref(query): Integrate readthrough cache

tkaemming opened the pull request:
This integrates the `get_readthrough` cache method introduced in #882 into the `execute_query` function stack (when enabled), circumventing `execute_query_with_deduplication` and `execute_query_with_caching`.

tkaemming committed:
structure

tkaemming committed:
integrate without stats

tkaemming committed:
include timings in get_readthrough

fpacifici commented on snuba/state/cache/abstract.py:
```diff
@@ -30,7 +32,7 @@ def set(self, key: str, value: TValue) -> None:
     @abstractmethod
     def get_readthrough(
-        self, key: str, function: Callable[[], TValue], timeout: int
+        self, key: str, function: Callable[[], TValue], timeout: int, timer: Timer,
```
What about making this optional? 
The timer in the end is used to track metrics, so it could be up to the caller whether they are actually interested in tracking metrics since we package this class as generic.

tkaemming replied:
Did this in 2f3bd0b.

lynnagara replied:
@fpacifici Are you also suggesting making timer optional throughout the entire codebase or just here? It's required everywhere else.

tkaemming committed:
Merge branch 'master' into query-execution-strategy

tkaemming committed:
make timer optional

tkaemming commented:
I still haven't found a way to expose whether or not a query is a duplicate in a clean way and that is consistent with the previous implementation. One option that I considered is changing this return value to `Tuple[bool, TValue]` where the first member denotes whether or not the query had to wait — but since a query may wait and _still throw_, this wouldn't cover all cases. 
This seems to imply that the cache backend needs to be able to mutate some state in the case of success _or_ failure (and that eventually gets passed to the query log as well as the response body), but we don't have any metrics abstractions that allow that.

This is currently covered in metrics (any query that has a `dedupe_wait` timing is a query that was blocked) but will need to be addressed differently at some point for the querylog. Since the query log is currently unused, I personally don't think we need to block on this.

I don't think passing in `stats: MutableMapping[str, Any]` is a good idea here, but I guess we can if we're absolutely insistent on maintaining compatibility. I also don't want to block this change on coming up with a better approach for dealing with these kind of stats (which still I think is something along the lines of https://github.com/getsentry/snuba/pull/840#discussion_r395384443 but will need a larger discussion, especially since this diverges somewhat from the current implementation querylog and stats in the response body.)

Overall, I really just want to see how this actually impacts production before continuing to invest in tangentially related changes before being able to deploy it.

fpacifici reviewed (approved):
I am ok with this. I don't see an easy way to extract is_dupe from the cache. 
Could you please mark on the querylog storage and querylog processor that the is_dupe field is going to be empty? I suspect it will stay that way for quite some time, at least we can make it clear it is not an oversight. 
Or, even better, just remove the fields from the querylog and add it to the backlog if it is not already there.

tkaemming committed:
Merge branch 'master' into query-execution-strategy

tkaemming committed:
add not eabout is_duplicate inaccuracy when using get_readthrough

tkaemming commented:
> Could you please mark on the querylog storage and querylog processor that the is_dupe field is going to be empty? I suspect it will stay that way for quite some time, at least we can make it clear it is not an oversight.
> Or, even better, just remove the fields from the querylog and add it to the backlog if it is not already there.

I added this as a comment to those locations and also added on the backlog. I don't have a strong feeling about whether or not it's removed from the storage schema entirely, but if we do that I think it's a significant enough change that it would make sense to do separately.
